lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,369 @@
+
+        """
+        self.allbuffers[self.i, :] = buffer
+        self.i += 1
+        return 0.0
+
+
+class LapseRate(object):
+    """Plugin to calculate the lapse rate from orography and temperature
+       cubes.
+
+    References:
+        The method applied here is based on the method used in the 2010 paper:
+        https://rmets.onlinelibrary.wiley.com/doi/abs/10.1002/met.177
+
+    Code methodology:
+    1) Apply land/sea mask to temperature and orography datasets. Mask sea
+       points as NaN since image processing module does not recognise Numpy
+       masks.
+    2) Extracts neighbourhoods from both datasets:
+       Apply "generic_filter" image processing module to temperature data
+       to extract a neighbourhood for each single point and save this
+       neighbourhood into a larger array.
+       Repeat for orography data.
+       generic_filter mode="constant" ensures that points beyond edges of
+       dataset will be filled with "cval". (NaN in this case)
+    3) For all the stored orography neighbourhoods - take the neighbours around
+       the central point and create a mask where the height difference from
+       the central point is greater than 35m.
+    4) Loop through array of neighbourhoods and take the height and temperature
+       of all grid points and calculate the
+       temperature/height gradient = lapse rate
+    5) Constrain the lapse rate as > DALR and < -3.0*DALR
+
+    """
+
+    def __init__(self, max_height_diff=35, nbhood_radius=1,
+                 max_lapse_rate=-3*DALR, min_lapse_rate=DALR):
+        """Initialise class.
+        Args:
+            max_height_diff (float):
+                Maximum allowable height difference between the central point
+                and points in the neighbourhood over which the lapse rate will
+                be calculated (metres).
+
+            nbhood_radius (int):
+                Radius of neighbourhood around each point. The neighbourhood
+                will be a square array with side length 2*nbhood_radius + 1.
+
+            max_lapse_rate (float):
+                Maximum lapse rate allowed.
+
+            min_lapse_rate (float):
+                Minimum lapse rate allowed.
+
+        """
+
+        self.max_height_diff = max_height_diff
+        self.max_lapse_rate = max_lapse_rate
+        self.min_lapse_rate = min_lapse_rate
+
+        # nbhood_size=3 corresponds to a 3x3 array centred on the
+        # central point.
+        self.nbhood_size = int((2*nbhood_radius) + 1)
+
+        # generic_filter extracts the neighbourhood and returns a 1D array.
+        # ind_central_point indicates where the central point would be on
+        # this array
+        self.nbhoodarray_size = self.nbhood_size**2
+        self.ind_central_point = int(self.nbhoodarray_size/2)
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        desc = ('<LapseRate>')
```

Comment:
There are lots of variables in the __init__ that should be printed out in the __repr__ method to give a description of this class as it has been initialised, e.g.
```<LapseRate: max_height_diff: 35, nbhood_radius: 1, max_lapse_rate: 0.0294, min_lapse_rate: -0.0098>```

Any __repr__ test will need to be updated equivalently.